### DISCUSSION TITLE
2FA for docmost

### DISCUSSION THREAD

**Discussion by proxforge** [docmost/docmost#150] [Ideas]
Would be nice to be able to secure the webinterface with 2FA.

**Comment by Node815** [comment]
I agree!  We all need this to be done!  This is a show stopper for me.

**Comment by psychoberries** [comment]
2FA would really make me feel more comfortable about securing the webinterface.

**Comment by LukasWillin** [comment]
I am currently looking into doc most. Having a general 2FA option like an OTP would be really nice. Or a simple way to implement it ourselves (I haven't looked at the source code yet, so you might be way ahead of me.)

**Reply by Vito0912** [reply]
I made a PR for TOTP (and OIDC) in branches of my fork. The TOTP branch will be rewritten to improve compatibility. TOTP will not be merged into the Docmost main project, see https://github.com/docmost/docmost/pull/1319, and will be part of the enterprise version (I am not sure if that works for you).

TLDR; I will maintain the fork below and eventually add TOTP. Maybe a option for some

My fork: https://github.com/Vito0912/forkmost (For TOTP: https://github.com/Vito0912/forkmost/pull/39 - I do not provide an image for it right now, so you will need to build it yourself. Please read the readme, as this fork adds many other things and also removes/restricts some things).

I need to test OIDC a bit more before I can finally rewrite and merge the TOTP branch into my project, so it will take some time. And this will definitely break codes that have already been set up with the current state of the fork, so you will need to add them again then. Because of this, I do not recommend using it now and that is also why I do not provide any image for the TOTP branch at the moment.

**Reply by LukasWillin** [reply]
I will absolutely have a look. Thank you so much! I have in fact looked at your code at the end of last week already (PR). Shame it wasn't accepted, but that's okay.
Not sure if I will get to properly review your version this week. Once I know more I would love to contact you again either for helping you out on your fork or whether we are allowed to fork from you again.

Rest assured, I very much appreciate your response!

**Reply by Vito0912** [reply]
I am currently rewriting TOTP for it to be MFA (/ supporting totp/mail and maybe Passkeys [but later are unfamiliar to me and I have to research a bit. So I focus on TOTP and Mail]), and I do not think I will finish this week because I want to make it a bit more general. At the same time, I need to debug the merged OIDC to make sure there are no problems. So, reviewing the TOTP code this week does not make much sense imo ;) - But you can the OIDC ;);).

About forking: I respect the license used and have not changed it for any parts I contributed. As long as your fork also follows the same license (which is one of the best imo, so thanks Phillipo for using it), you are free to fork it again and make your own version. Of course, I would appreciate contributions, but that is completely up to you. (And a fun fact: Any code I made and had as a PR to this upstream project cannot be restricted by me, no matter what the project license is due to the CLA.)
